### PR TITLE
Added support for loop_entry history variable.

### PR DESCRIPTION
**Pull request by aalok-thakkar**
In this commit we introduce a new expression, `__CPROVER_loop_entry`, which can be used within loop invariants to track history of variables.
`__CPROVER_loop_entry(x)` returns the initial value of a variable just before the very first iteration of a loop.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- n/a ~~White-space or formatting changes outside the feature-related changed lines are in commits of their own.~~

### CODE REVIEW

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: Why is this better than the previous message?

**Review by feliperodri — CHANGES_REQUESTED** [review]
We need regression tests.

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: I think Aalok was trying to generalize the error message for both `__CPROVER_old` and `__CPROVER_loop_entry`.

I just have a minor suggestion:
```suggestion
      expr.operands().size() == 1, "A history variable must have one operand.");
```

Note that this is not a user-facing error message, but an internal invariant failure in case we tried to construct a bad history variable. But we can improve the invariant failure message, if this is unclear. Do you have any suggestions, @feliperodri?

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: Do we need to split the line here?

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]:
```suggestion
      log.error() << "History variables do not currently support "
```

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: May be you meant:
```suggestion
  insert_before_swap_and_advance(goto_function.body, loop_head, history);
```
because `loop_head` is an instruction iterator into that, not `havoc_code`?

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: As per our offline discussion, may be just reuse `havoc_code` instead of defining this?

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]

File: regression/contracts/loop-history-variables-01/main.c
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-01/main.c]: Please add a newline at the end (this file and others).

File: regression/contracts/loop-history-variables-01/main.c
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-01/main.c]: Could you also add a couple more loops to this file and test for struct members and scalars as well?

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]

File: regression/contracts/loop-history-variables-01/test.desc
Comment (feliperodri) [inline regression/contracts/loop-history-variables-01/test.desc]: Add new line.

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: @SaswatPadhi if we add info of the source location, we could perhaps avoid code duplication and keep a single error message for both history variables. What do you think? Only a suggestion.

**Review by feliperodri — APPROVED** [review]
LGTM modulo Saswat's comments.

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.cpp]: See https://github.com/diffblue/cbmc/pull/6278#discussion_r684582662

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]:
```suggestion
      log.error() << "History variables do not currently support "
                  << parameter.id() << " expressions." << messaget::eom;
      throw 0;
```

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: This is an internal invariant (`DATA_INVARIANT`) failure, not a user-facing error message; so I don't think source location would make much sense.

But looking at this code now, I don't think this entire thing is necessary at all. You can remove lines 404 - 414.

The `to_old_expr` call @ line 415 would call `validate_expr` for the base class `unary_expr`, which would anyway check the number of operands internally:

https://github.com/diffblue/cbmc/blob/5b8e8fa6fc8eb6b241aaca38e4e706568a7ac158/src/util/std_expr.h#L317-L320

If we want to show user-facing errors, that is already done within the parser.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]

File: src/goto-instrument/contracts/contracts.cpp
Comment (aalok-thakkar) [reply src/goto-instrument/contracts/contracts.cpp]: I've made separate functions and avoided generalization.

Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: Please don't do that. That looks like a lot of duplicate code.

As I have commented below, this operand check can simply be removed.

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: The 2 lines above are redundant. Please remove those.

The number of operands check would be performed during `to_history_expr` call @ line 411, as I mentioned in https://github.com/diffblue/cbmc/pull/6278#discussion_r684686555.

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: Not addressed yet.

File: src/ansi-c/c_typecheck_code.cpp
Comment (SaswatPadhi) [inline src/ansi-c/c_typecheck_code.cpp]: Why do we need this? There is `c_typecheck_baset::disallow_history_variables` defined in `c_typecheck_expr.cpp`

File: regression/contracts/loop-history-variables-04/test.desc
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-04/test.desc]: 1. There are no `struct`s in the test case.
2. We are introducing too many tiny regression test files. Let's just have a single C file with multiple loops, each one checking a different kind of objects. The line number of the errors (whenever that happens) would indicate the root cause; we don't need separate files.

File: regression/contracts/loop-history-variables-03/test.desc
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-03/test.desc]: But we are not checking that :) The C file is using `__CPROVER_old` in function contracts.

File: regression/contracts/loop-history-variables-03/main.c
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-03/main.c]: `oid`?!

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]

File: src/ansi-c/c_typecheck_expr.cpp
Comment (feliperodri) [inline src/ansi-c/c_typecheck_expr.cpp]: Are not allowed where? Could you improve this error message?

File: regression/contracts/history-pointer-replace-03/test.desc
Comment (feliperodri) [inline regression/contracts/history-pointer-replace-03/test.desc]: Not allowed where? Could you improve this error message?

File: regression/contracts/loop-history-variables-01/main.c
Comment (feliperodri) [inline regression/contracts/loop-history-variables-01/main.c]: You need stdlib for malloc too.

File: regression/contracts/loop-history-variables-02/test.desc
Comment (feliperodri) [inline regression/contracts/loop-history-variables-02/test.desc]: Which error message shall we get in this case?

File: regression/contracts/loop-history-variables-03/test.desc
Comment (feliperodri) [inline regression/contracts/loop-history-variables-03/test.desc]: Which error message shall we get in this case?

**Review by feliperodri — CHANGES_REQUESTED** [review]
Almost there 😊

File: regression/contracts/history-pointer-replace-03/test.desc
Comment (SaswatPadhi) [reply regression/contracts/history-pointer-replace-03/test.desc]: The line number is printed on the standard output, but we don't match it in the regex. Should we match the line number here?

File: src/ansi-c/c_typecheck_expr.cpp
Comment (SaswatPadhi) [reply src/ansi-c/c_typecheck_expr.cpp]: Since it's a common error message for disallowing `__CPROVER_old` within requires clauses and loop invariants, we shouldn't say "in __CPROVER_requires clauses" any more. May be we could say:

```
 error: __CPROVER_old expressions are not allowed in this context
```

Since we set the "source_location", the line number should also be printed and "in this context" would be clear from that.

**Review by SaswatPadhi — COMMENTED** [review]
There is a bit of code duplication between `disallow_loop_history_variables` and `disallow_history_variables`.

Can we remove `disallow_loop_history_variables`, add an additional argument to `disallow_history_variables` which would be the `id` of the history variable expressions we want to disallow, and just reuse the code in both contexts?

**Review by SaswatPadhi — APPROVED** [review]
Thanks; LGTM modulo the build issues with CI.

**Comment by feliperodri** [comment]
@aalok-thakkar CI failures are due to regex on test descriptions (not compatible with Windows). You also need to run clang-format.

File: regression/contracts/history-pointer-replace-03/test.desc
Comment (feliperodri) [inline regression/contracts/history-pointer-replace-03/test.desc]: Not compatible with Windows system.

File: regression/contracts/loop-history-variables-02/test.desc
Comment (feliperodri) [inline regression/contracts/loop-history-variables-02/test.desc]: Not compatible with Windows system.

File: regression/contracts/loop-history-variables-03/test.desc
Comment (feliperodri) [inline regression/contracts/loop-history-variables-03/test.desc]: Not compatible with Windows system.

**Review by feliperodri — APPROVED** [review]
LGTM modulo CI failures.

**Review by feliperodri — APPROVED** [review]

**Review by chrisr-diffblue — COMMENTED** [review]
Code looks great, but I wonder if I might ask for a couple of extras:

1) Some more negative tests - at the moment the regression tests you've added all lead either to VERIFICATION SUCCESSFUL, or CONVERSION ERROR... Could we perhaps have some tests that also lead to verification failure? This might also help with my next point :-)
2) This is more optional... but it would be _super_ nice if we could also have a small bit of user documentation for this new predicate... indicating its purpose, how it should be used, etc.

I won't block the review just for missing user documentation, but I think we could do with the extra "negative test" before we merge.

File: regression/contracts/loop-history-variables-04/test.desc
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-04/test.desc]: How exactly are we checking that `__CPROVER_loop_entry` is supported? It doesn't appear in the the C program ...
The parsing test cases already check for support. The negative tests should be checking for assertion failures (see my comment above).

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Thanks @chrisr-diffblue!

@aalok-thakkar I don't understand the new test. I think by negative tests Chris was suggesting that we include a test where it would be violated, and we check that CBMC correctly catches it. An example would be:

```c
while(y > 0)
    __CPROVER_loop_invariant(x == __CPROVER_loop_entry(x))
    {
      --y;
      x += 1;
      x -= 2;
    }
}
```

**Review by SaswatPadhi — APPROVED** [review]
LGTM. Thanks for all the fixes, @aalok-thakkar :)

**Comment by feliperodri** [comment]
@chrisr-diffblue could you take another look?

File: regression/contracts/loop-history-variables-01/main.c
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-01/main.c]:
```suggestion
  x1 = &z1;
```

File: regression/contracts/loop-history-variables-02/main.c
Comment (SaswatPadhi) [inline regression/contracts/loop-history-variables-02/main.c]:
```suggestion
  x = &z;
```

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Some issues with pointers.

(They were not caught because the tests are not run with `--pointer-check`)

**Comment by SaswatPadhi** [comment]
The CI failures are due to `.desc` files being out of sync. You should also update `*x` -> `x` in those files.

Minor comment on the commit description: please change `a new predicate __CPROVER_loop_entry` -> `a new unary expression __CPROVER_loop_entry`. `__CPROVER_loop_entry` is not a predicate (a `bool`-returning function).

**Review by chrisr-diffblue — APPROVED** [review]
Approved, pending fixing the issue @SaswatPadhi has mentioned with --pointer-check

**Review by SaswatPadhi — APPROVED** [review]

**Pull request merged**
